### PR TITLE
task/RFR-225_refactor-config

### PR DESCRIPTION
**Pull request by muthenberg**
This is a more idiomatic way of providing the configuration. I was unable to get POJO parsing to work, so I wrote my own parser. This should also reduce the dependency on the API project. I believe sharing settings between provider and collector there is not the best solution, since those settings needn't to be equal.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -103,7 +103,7 @@ tasks.withType<KotlinCompile>().configureEach {
 }
 
 // If you increase this version, check in the next line if the manual mongo driver version is still necessary.
-extra["vertxVersion"] = "4.3.5"
+extra["vertxVersion"] = "4.3.6"
 // The following is only required since Vert.x GridFS Client is not working correctly in Version 4.3.3.
 // To check this run GridFSStorageIT
 // We reported the problem to Vertx Github. A fix is scheduled for Vertx 4.3.4
```
Comment (hb0) [inline build.gradle.kts:109]: Shouldn't we check if we can remove the specific `mongoDriverVersion` as the comment suggests?

File: src/main/kotlin/de/cyface/collector/configuration/Configuration.kt [hunk 2]
```diff
@@ -136,6 +177,13 @@ data class Configuration(
             }
         }
 
+        // TODO [2022-12-14] Klemens Muthmann: This code is ported from the previous implementation.
+        // Armin Schnabel has added the original code, but I do not really understand why it is necessary.
+        // The documentation should be improved when I understand why it is necessary to add a * at the end of the
+        // endpoint.
+        /**
+         * Provide the HTTP endpoint configured for this data collector.
+         */
```
Comment (hb0) [inline src/main/kotlin/de/cyface/collector/configuration/Configuration.kt:186]: I think this was added [here](https://github.com/cyface-de/data-collector/commit/d10550908e50fc2b161ccba8d5178d33d71257df) by you, it seems like this was a requirement in the upgraded Vert.X version?

Comment (hb0) [reply src/main/kotlin/de/cyface/collector/configuration/Configuration.kt:186]: As we use `httpEndpoint() { return "api/v3/*" }` in `mainRouter.route(config.getEndpoint()).subRouter(apiRouter);` I guess you have this requirement from [here](https://vertx.io/docs/vertx-web/java/#_routing_by_paths_that_begin_with_something) in the Vert.X documentation.

I.e. we route everything at `api/v3/*****` to the apiV3Router which handles sub-routes.

And everything else by the failure handler:
`mainRouter.route("/*").last().handler(new de.cyface.api.FailureHandler());`

File: src/test/kotlin/de/cyface/collector/storage/cloud/MetaDataStorageIT.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+        assertTrue(lock.await(10, TimeUnit.SECONDS))
+    }
+
+    /**
+     * Provide a working client configuration. Note that you need to change username and password for this to work.
+     */
+    private fun configuration(): MongoClientSettings {
+        val connString = ConnectionString(
+            "mongodb+srv://<username>:<password>@cyface.o1bn5yx.mongodb.net/?retryWrites=true&w=majority"
```
Comment (hb0) [inline src/test/kotlin/de/cyface/collector/storage/cloud/MetaDataStorageIT.kt:74]: Is this hostname static on purpose?

File: conf.json.template [hunk 4]
```diff
@@ -13,5 +13,13 @@
 	"http.port":8080,
 	"http.host":"localhost",
 	"http.endpoint":"/api/v3",
-	"http.port.management":13371
+	"http.port.management":13371,
+	"salt": "cyface_salt",
+	"upload.expiration": 600,
+	"measurement.payload.limit": 100,
```
Comment (hb0) [inline conf.json.template:19]: Are you sure this value is correct?
In the old code the default value for 100 MB was:

```
/**
     * The default number of seconds the JWT authentication token is valid after login.
     */
    private static final int DEFAULT_PAYLOAD_LIMIT = Math.toIntExact(BYTES_IN_ONE_MEGABYTE * 100L);
```

With `BYTES_IN_ONE_MEGABYTE = 1_048_576L`

File: src/main/docker/entrypoint.sh [hunk 5]
```diff
@@ -122,7 +122,15 @@ loadConfig() {
       \"http.port.management\":$CYFACE_MANAGEMENT_PORT,\
       \"admin.user\":\"$ADMIN_USER\",\
       \"admin.password\":\"$ADMIN_PASSWORD\",\
-      \"metrics.enabled\":$METRICS_ENABLED\
+      \"metrics.enabled\":$METRICS_ENABLED,\
+      \"salt\":\"cyface-salt\",\
+      \"upload.expiration\":60,\
+      \"measurement.payload.limit\":100,\
```
Comment (hb0) [inline src/main/docker/entrypoint.sh:128]: should be `100 * (BYTES_IN_ONE_MEGABYTE = 1_048_576L)` ?

File: src/main/kotlin/de/cyface/collector/configuration/Configuration.kt [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+ * @property jwtPublic The [Path] to the public key file used for JWT authentication.
+ * @property serviceHttpAddress The [URL] hosting the collector service.
+ * @property mongoDb The Mongo database configuration as described in the Vert.x documentation.
+ * @property adminUser The name of the default admin user to create on startup if it does not exist yet.
+ * @property adminPassword The password for the default user to create if it does not exist. This should be changed in
+ * production.
+ * @property salt The salt to make breaking passwords harder.
+ * @property jwtExpiration The time a JWT token stays valid.
+ * @property uploadExpiration The time an upload session stays valid to be resumed in the future.
+ * @property measurementPayloadLimit The maximum size accepted for a single measurement.
```
Comment (hb0) [inline src/main/kotlin/de/cyface/collector/configuration/Configuration.kt:45]: should probably add the unit: bytes

File: src/test/java/de/cyface/collector/verticle/CollectorApiVerticleTest.java [hunk 7]
```diff
@@ -86,32 +91,34 @@ void shutdown() {
...
+                        "localhost",
+                        port,
+                        "/api/v3/*"
+                ));
+        when(configuration.getMongoDb()).thenReturn(mongoTest.clientConfiguration());
+        when(configuration.getJwtExpiration()).thenReturn(3600);
+        when(configuration.getSalt()).thenReturn(new ValueSalt("cyface-salt"));
+        when(configuration.getStorageType()).thenReturn(new GridFsStorageType(Path.of("upload-folder")));
+        when(configuration.getUploadExpiration()).thenReturn(60_000L);
+        when(configuration.getMeasurementPayloadLimit()).thenReturn(100L);
```
Comment (hb0) [inline src/test/java/de/cyface/collector/verticle/CollectorApiVerticleTest.java:113]: should be `100 * (BYTES_IN_ONE_MEGABYTE = 1_048_576L)` ?

File: src/test/java/de/cyface/collector/verticle/MainVerticleTest.java [hunk 8]
```diff
@@ -100,7 +100,15 @@ private JsonObject config() throws IOException {
                 .put("http.endpoint", "/api/v3/")
                 .put("http.port", Network.freeServerPort(Network.getLocalHost()))
                 .put("salt", "abcdefg")
-                .put("mongo.db", mongoTest.clientConfiguration());
+                .put("mongo.db", mongoTest.clientConfiguration())
+                .put("admin.user", "admin")
+                .put("admin.password", "secret")
+                .put("jwt.expiration", 60)
+                .put("upload.expiration", 60)
+                .put("measurement.payload.limit", 100)
```
Comment (hb0) [inline src/test/java/de/cyface/collector/verticle/MainVerticleTest.java:108]: should be `100 * (BYTES_IN_ONE_MEGABYTE = 1_048_576L)` ?

File: src/test/resources/configurations/test-conf-missing.json [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    "db_name": "cyface",
+    "connection_string": "mongodb+srv://user:password@mongodb.net/?retryWrites=true&w=majority",
+    "data_source_name": "cyface"
+  },
+  "admin.user": "admin",
+  "admin.password": "secret",
+  "salt": "cyface_salt",
+  "jwt.expiration": 60,
+  "upload.expiration": 60,
+  "measurement.payload.limit": 100,
```
Comment (hb0) [inline src/test/resources/configurations/test-conf-missing.json:16]: should be `100 * (BYTES_IN_ONE_MEGABYTE = 1_048_576L)` ?

File: src/test/resources/configurations/test-conf-with-additional-parameter.json [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    "db_name": "cyface",
+    "connection_string": "mongodb+srv://user:password@mongodb.net/?retryWrites=true&w=majority",
+    "data_source_name": "cyface"
+  },
+  "admin.user": "admin",
+  "admin.password": "secret",
+  "salt": "cyface_salt",
+  "jwt.expiration": 60,
+  "upload.expiration": 60,
+  "measurement.payload.limit": 100,
```
Comment (hb0) [inline src/test/resources/configurations/test-conf-with-additional-parameter.json:18]: should be `100 * (BYTES_IN_ONE_MEGABYTE = 1_048_576L)` ?

File: src/test/resources/configurations/test-conf-with-salt-and-saltpath.json [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    "connection_string": "mongodb+srv://user:password@mongodb.net/?retryWrites=true&w=majority",
+    "data_source_name": "cyface"
+  },
+  "admin.user": "admin",
+  "admin.password": "secret",
+  "salt": "cyface_salt",
+  "salt.path": "path/to/salt/file",
+  "jwt.expiration": 60,
+  "upload.expiration": 60,
+  "measurement.payload.limit": 100,
```
Comment (hb0) [inline src/test/resources/configurations/test-conf-with-salt-and-saltpath.json:18]: should be `100 * (BYTES_IN_ONE_MEGABYTE = 1_048_576L)` ?

File: src/test/resources/configurations/test-conf.json [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    "db_name": "cyface",
+    "connection_string": "mongodb+srv://user:password@mongodb.net/?retryWrites=true&w=majority",
+    "data_source_name": "cyface"
+  },
+  "admin.user": "admin",
+  "admin.password": "secret",
+  "salt": "cyface_salt",
+  "jwt.expiration": 60,
+  "upload.expiration": 60,
+  "measurement.payload.limit": 100,
```
Comment (hb0) [inline src/test/resources/configurations/test-conf.json:17]: should be `100 * (BYTES_IN_ONE_MEGABYTE = 1_048_576L)` ?

File: conf.json.template [hunk 13]
```diff
@@ -13,5 +13,13 @@
 	"http.port":8080,
 	"http.host":"localhost",
 	"http.endpoint":"/api/v3",
-	"http.port.management":13371
+	"http.port.management":13371,
+	"salt": "cyface_salt",
+	"upload.expiration": 600,
```
Comment (hb0) [inline conf.json.template:18]: You probably meant 10 minutes, i.e. `1000 * 60 * 10` ?

Why did you change the default upload expiration time of 1 week?

```
/**
     * The default number of milliseconds to wait before removing cached uploads after last modification: 7 days.
     */
    private static final long DEFAULT_UPLOAD_EXPIRATION_TIME = 1000 * 60 * 60 * 24 * 7;
```

File: src/main/docker/entrypoint.sh [hunk 14]
```diff
@@ -122,7 +122,15 @@ loadConfig() {
       \"http.port.management\":$CYFACE_MANAGEMENT_PORT,\
       \"admin.user\":\"$ADMIN_USER\",\
       \"admin.password\":\"$ADMIN_PASSWORD\",\
-      \"metrics.enabled\":$METRICS_ENABLED\
+      \"metrics.enabled\":$METRICS_ENABLED,\
+      \"salt\":\"cyface-salt\",\
+      \"upload.expiration\":60,\
```
Comment (hb0) [inline src/main/docker/entrypoint.sh:127]: You probably meant 1 minute, i.e. `1000 * 60 * 1` ?

Why did you change the default upload expiration time of 1 week?

File: src/main/kotlin/de/cyface/collector/configuration/Configuration.kt [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+ * @property jwtPrivate The [Path] to the private key file used for JWT authentication.
+ * @property jwtPublic The [Path] to the public key file used for JWT authentication.
+ * @property serviceHttpAddress The [URL] hosting the collector service.
+ * @property mongoDb The Mongo database configuration as described in the Vert.x documentation.
+ * @property adminUser The name of the default admin user to create on startup if it does not exist yet.
+ * @property adminPassword The password for the default user to create if it does not exist. This should be changed in
+ * production.
+ * @property salt The salt to make breaking passwords harder.
+ * @property jwtExpiration The time a JWT token stays valid.
+ * @property uploadExpiration The time an upload session stays valid to be resumed in the future.
```
Comment (hb0) [inline src/main/kotlin/de/cyface/collector/configuration/Configuration.kt:44]: we might want to add the units here:

- jwtExpiration = seconds (int)
- uploadExpiration = milliseconds (long)

File: src/test/java/de/cyface/collector/verticle/MainVerticleTest.java [hunk 16]
```diff
@@ -100,7 +100,15 @@ private JsonObject config() throws IOException {
                 .put("http.endpoint", "/api/v3/")
                 .put("http.port", Network.freeServerPort(Network.getLocalHost()))
                 .put("salt", "abcdefg")
-                .put("mongo.db", mongoTest.clientConfiguration());
+                .put("mongo.db", mongoTest.clientConfiguration())
+                .put("admin.user", "admin")
+                .put("admin.password", "secret")
+                .put("jwt.expiration", 60)
+                .put("upload.expiration", 60)
```
Comment (hb0) [inline src/test/java/de/cyface/collector/verticle/MainVerticleTest.java:107]: `jwt.expiration  = 60` is correct (seconds)

but `upload.expiration = 60` should probably be 60_000 (ms)

File: src/test/kotlin/de/cyface/collector/commons/ConfigurationFactory.kt [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+                URL(
+                    "https",
+                    "localhost",
+                    port,
+                    "/api/v3/*"
+                )
+            on { adminUser } doReturn "admin"
+            on { adminPassword } doReturn "secret"
+            on { salt } doReturn ValueSalt("cyface-salt")
+            on { jwtExpiration } doReturn 10_000
```
Comment (hb0) [inline src/test/kotlin/de/cyface/collector/commons/ConfigurationFactory.kt:62]: should probably be 10 not 10_000 (unit: seconds not ms)

File: src/test/resources/configurations/test-conf-missing.json [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+  "mongo.db": {
+    "db_name": "cyface",
+    "connection_string": "mongodb+srv://user:password@mongodb.net/?retryWrites=true&w=majority",
+    "data_source_name": "cyface"
+  },
+  "admin.user": "admin",
+  "admin.password": "secret",
+  "salt": "cyface_salt",
+  "jwt.expiration": 60,
+  "upload.expiration": 60,
```
Comment (hb0) [inline src/test/resources/configurations/test-conf-missing.json:15]: You probably meant `1000 * 60 * 1` ?

File: src/test/resources/configurations/test-conf-with-additional-parameter.json [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+  "mongo.db": {
+    "db_name": "cyface",
+    "connection_string": "mongodb+srv://user:password@mongodb.net/?retryWrites=true&w=majority",
+    "data_source_name": "cyface"
+  },
+  "admin.user": "admin",
+  "admin.password": "secret",
+  "salt": "cyface_salt",
+  "jwt.expiration": 60,
+  "upload.expiration": 60,
```
Comment (hb0) [inline src/test/resources/configurations/test-conf-with-additional-parameter.json:17]: You probably meant `1000 * 60 * 1` ?

File: src/test/resources/configurations/test-conf-with-salt-and-saltpath.json [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+    "db_name": "cyface",
+    "connection_string": "mongodb+srv://user:password@mongodb.net/?retryWrites=true&w=majority",
+    "data_source_name": "cyface"
+  },
+  "admin.user": "admin",
+  "admin.password": "secret",
+  "salt": "cyface_salt",
+  "salt.path": "path/to/salt/file",
+  "jwt.expiration": 60,
+  "upload.expiration": 60,
```
Comment (hb0) [inline src/test/resources/configurations/test-conf-with-salt-and-saltpath.json:17]: You probably meant `1000 * 60 * 1` ?

File: src/test/resources/configurations/test-conf.json [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+  "mongo.db": {
+    "db_name": "cyface",
+    "connection_string": "mongodb+srv://user:password@mongodb.net/?retryWrites=true&w=majority",
+    "data_source_name": "cyface"
+  },
+  "admin.user": "admin",
+  "admin.password": "secret",
+  "salt": "cyface_salt",
+  "jwt.expiration": 60,
+  "upload.expiration": 60,
```
Comment (hb0) [inline src/test/resources/configurations/test-conf.json:16]: You probably meant `1000 * 60 * 1` ?

File: README.md [hunk 22]
```diff
@@ -94,39 +100,59 @@ Use `docker-compose ps` to see which ports are mapped to which by Docker.
...
+* **salt.path:** The path to a salt file used to encrypt passwords stored in the user database even stronger.
+* **salt:** A salt value that may be used instead of the salt from salt.path. You must make sure that either the salt or the salt.path parameter are used. If both are specified the application startup will fail.
 * **metrics.enabled:** Set to either `true` or `false`. If `true` the collector API publishes metrics using micrometer. These metrics are accessible by a [Prometheus](https://prometheus.io/) server (Which you need to set up yourself) at port `8081`.
+* **http.port.management:** The port running the management API responsible for creating user accounts.
+* **jwt.expiration**: The time it takes for a JWT token to expire in milliseconds. If a JWT token expires, clients need to acquire a new one via username and password authentication. Setting this time too short requires sending the username and password more often. This makes it easier for malicious parties to intercept and brute force usernames and passwords. However long time JWT tokens may be captured as well and used for malicious purposes.
```
Comment (hb0) [inline README.md:143]: Wrong unit, this is currently an INT in seconds.

**Review by hb0 — CHANGES_REQUESTED** [review]
See comments

File: build.gradle.kts (hunk 1 above)
Comment (muthenberg) [reply build.gradle.kts:109]: Did so. Does not work yet.

File: conf.json.template (hunk 4 above)
Comment (muthenberg) [reply conf.json.template:19]: This is just some example value and needs to be changed. The problem is that I did not find documentation on whether this is supposed to be in Bytes, Kilotbytes or Megabytes. But if it is Bytes it should probably be changed.

File: src/test/resources/configurations/test-conf-missing.json (hunk 18 above)
Comment (muthenberg) [reply src/test/resources/configurations/test-conf-missing.json:15]: No. I just required a value here. These files are just used to test parsing, not to really start a server.

File: src/test/kotlin/de/cyface/collector/storage/cloud/MetaDataStorageIT.kt (hunk 3 above)
Comment (muthenberg) [reply src/test/kotlin/de/cyface/collector/storage/cloud/MetaDataStorageIT.kt:74]: It is a placeholder for a hostname to a real Mongo DB Atlas Cluster, which looks similar. But we can probably use something more generic.

**Review by hb0 — APPROVED** [review]

File: src/test/resources/configurations/test-conf-missing.json (hunk 18 above)
Comment (hb0) [reply src/test/resources/configurations/test-conf-missing.json]: I would just be confused by "60" for milliseconds ;-) 1 or 1_000 or 60_000 is better, thanks for changing that :)
